api/src/main/java/ca/bc/gov/educ/api/gradstudent/service/GradStudentPaginationService.java

```diff
@@ -0,0 +1,153 @@
+        return results;
+      } catch (final Throwable ex) {
+        log.error("Failure querying for paginated SDC school students: {}", ex.getMessage());
+        throw new CompletionException(ex);
+      }
+    }, paginatedQueryExecutor);
+
+  }
+
+  public Specification<GraduationStudentRecordPaginationEntity> setSpecificationAndSortCriteria(String sortCriteriaJson, String searchCriteriaListJson, ObjectMapper objectMapper, List<Sort.Order> sorts) {
+    Specification<GraduationStudentRecordPaginationEntity> schoolSpecs = null;
+    try {
+      RequestUtil.getSortCriteria(sortCriteriaJson, objectMapper, sorts);
+      if (StringUtils.isNotBlank(searchCriteriaListJson)) {
+        List<Search> searches = objectMapper.readValue(searchCriteriaListJson, new TypeReference<>() {
+        });
+        int i = 0;
+        for (var search : searches) {
+          schoolSpecs = getSpecifications(schoolSpecs, i, search, gradStudentPaginationFilterSpecs);
+          i++;
+        }
+      }
+    } catch (JsonProcessingException e) {
+      throw new GradStudentAPIRuntimeException(e.getMessage());
+    }
+    return schoolSpecs;
+  }
+
+  public <T> Specification<T> getSpecifications(Specification<T> specs, int i, Search search, BaseFilterSpecs<T> filterSpecs) {
+    if (i == 0) {
+      specs = getEntitySpecification(search.getSearchCriteriaList(), filterSpecs);
+    } else {
+      if (search.getCondition() == Condition.AND) {
+        specs = specs.and(getEntitySpecification(search.getSearchCriteriaList(), filterSpecs));
+      } else {
+        specs = specs.or(getEntitySpecification(search.getSearchCriteriaList(), filterSpecs));
+      }
+    }
+    return specs;
+  }
+
+  private <T> Specification<T> getEntitySpecification(List<SearchCriteria> criteriaList, BaseFilterSpecs<T> filterSpecs) {
+    Specification<T> specs = null;
+    if (!criteriaList.isEmpty()) {
+      int i = 0;
+      for (SearchCriteria criteria : criteriaList) {
+        if (criteria.getKey() != null && criteria.getOperation() != null && criteria.getValueType() != null) {
+          var criteriaValue = criteria.getValue();
+          if(StringUtils.isNotBlank(criteria.getValue()) && TransformUtil.isUppercaseField(GraduationStudentRecordPaginationEntity.class, criteria.getKey())) {
+            criteriaValue = criteriaValue.toUpperCase();
+          }
+          Specification<T> typeSpecification = getTypeSpecification(criteria.getKey(), criteria.getOperation(), criteriaValue, criteria.getValueType(), filterSpecs);
+          specs = getSpecificationPerGroup(specs, i, criteria, typeSpecification);
+          i++;
+        } else {
+          throw new InvalidParameterException("Search Criteria can not contain null values for key, value and operation type");
+        }
+      }
+    }
+    return specs;
+  }
+
+  private <T> Specification<T> getSpecificationPerGroup(Specification<T> entitySpecification, int i, SearchCriteria criteria, Specification<T> typeSpecification) {
```

Comment:
Move to parent class or utility class (also used in ReportGradStudentSearchService with generics)